File: CS/SilverlightApplication41/NWindData.cs
using System;
using System.Collections.Generic;
using System.Linq;
using SilverlightApplication41.Web;
using System.Threading;
using System.ServiceModel.DomainServices.Client;

namespace SilverlightApplication41 {
    public static class NWindData {
        static EventHandler loaded;
        static bool loadingDone;
        static AutoResetEvent loadingDoneUnlock = new AutoResetEvent(true); // mutex, protect loadingDone and loaded
        static bool loadingStarted;
        static AutoResetEvent loadingStartedUnlock = new AutoResetEvent(true); // mutex, protect loadingStarted
        static NWindContext dc = new NWindContext();
        static int tablesCount = 0;
        static int tablesLoaded = 0;
        public static void Load() {
            loadingStartedUnlock.WaitOne();
            if (!loadingStarted) {
                loadingStarted = true;
                Thread thread = new Thread(new ThreadStart(StartLoading));
                thread.Start();
            }
            loadingStartedUnlock.Set();
            loadingDoneUnlock.WaitOne();
            if (loadingDone) {
                if (loaded != null) loaded(typeof(NWindData), EventArgs.Empty);
                loaded = null;
            }
            loadingDoneUnlock.Set();
        }
        private static void StartLoading() {
            tablesCount = 1;
            dc.Load(dc.GetProductsQuery(), new Action<LoadOperation<Product>>(TLoaded), null);
        }
        public static event EventHandler Loaded {
            add {
                loadingDoneUnlock.WaitOne();
                loaded += value;
                loadingDoneUnlock.Set();
            }
            remove {
                loadingDoneUnlock.WaitOne();
                loaded -= value;
                loadingDoneUnlock.Set();
            }
        }

        static void TLoaded(LoadOperation<Product> op) {
            TableLoaded();
        }
        private static void TableLoaded() {
            if (++tablesLoaded >= tablesCount) {
                loadingDoneUnlock.WaitOne();
                loadingDone = true;
                if (loaded != null) {
                    loaded(typeof(NWindData), EventArgs.Empty);
                    loaded = null;
                }
                loadingDoneUnlock.Set();
            }
        }
        public static EntitySet<Product> Products { get { return dc.Products; } }
    }
}

File: VB/SilverlightApplication41/NWindData.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports SilverlightApplication41.Web
Imports System.Threading
Imports System.ServiceModel.DomainServices.Client

Namespace SilverlightApplication41
	Public NotInheritable Class NWindData
		Private Shared Event loaded As EventHandler
		Private Shared loadingDone As Boolean
		Private Shared loadingDoneUnlock As New AutoResetEvent(True) ' mutex, protect loadingDone and loaded
		Private Shared loadingStarted As Boolean
		Private Shared loadingStartedUnlock As New AutoResetEvent(True) ' mutex, protect loadingStarted
		Private Shared dc As New NWindContext()
		Private Shared tablesCount As Integer = 0
		Private Shared tablesLoaded As Integer = 0
		Private Sub New()
		End Sub
		Public Shared Sub Load()
			loadingStartedUnlock.WaitOne()
			If (Not loadingStarted) Then
				loadingStarted = True
				Dim thread As New Thread(New ThreadStart(AddressOf StartLoading))
				thread.Start()
			End If
			loadingStartedUnlock.Set()
			loadingDoneUnlock.WaitOne()
			If loadingDone Then
				RaiseEvent loaded(GetType(NWindData), EventArgs.Empty)
				loadedEvent = Nothing
			End If
			loadingDoneUnlock.Set()
		End Sub
		Private Shared Sub StartLoading()
			tablesCount = 1
			dc.Load(dc.GetProductsQuery(), New Action(Of LoadOperation(Of Product))(AddressOf TLoaded), Nothing)
		End Sub
		Public Shared Custom Event LoadedEv As EventHandler
			AddHandler(ByVal value As EventHandler)
				loadingDoneUnlock.WaitOne()
				AddHandler loaded, value
				loadingDoneUnlock.Set()
			End AddHandler
			RemoveHandler(ByVal value As EventHandler)
				loadingDoneUnlock.WaitOne()
				RemoveHandler loaded, value
				loadingDoneUnlock.Set()
			End RemoveHandler
			RaiseEvent(ByVal sender As System.Object, ByVal e As System.EventArgs)
			End RaiseEvent
		End Event

		Private Shared Sub TLoaded(ByVal op As LoadOperation(Of Product))
			TableLoaded()
		End Sub
		Private Shared Sub TableLoaded()
'INSTANT VB TODO TASK: Assignments within expressions are not supported in VB.NET
'ORIGINAL LINE: if (++tablesLoaded >= tablesCount)
			If ++tablesLoaded >= tablesCount Then
				loadingDoneUnlock.WaitOne()
				loadingDone = True
				If loadedEvent IsNot Nothing Then
					RaiseEvent loaded(GetType(NWindData), EventArgs.Empty)
					loadedEvent = Nothing
				End If
				loadingDoneUnlock.Set()
			End If
		End Sub
		Public Shared ReadOnly Property Products() As EntitySet(Of Product)
			Get
				Return dc.Products
			End Get
		End Property
	End Class
End Namespace
